### ISSUE TITLE
yield expressions should be able to acquire the type of generator return types

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#10509]
Courtesy of @bterlson (and his idea in general)

``` ts
interface Foo<T> extends IterableIterator<T> {
    next(valueGivenWhenYieldExpressionIsUsed: number)
}
```

This type is basically an `IterableIterator` _except_ that it _restricts_ the types of values a consumer can pass through to `next`.

``` ts
function* foo(): Foo<string> {
    let x = yield "hello";
    console.log(x);
}

let gen = foo();
gen.next(100);
gen.next(200);
```

Here, calls to `next` can _only_ accept `number`s; however, the yield expressions in the generator are just typed as `any`.

Since we have an explicit type, we should just figure the type of `yield` expressions out from that.

**Comment by DanielRosenwasser** [comment]
Marking as "Needs Proposal" because it's not entirely obvious what to do in the face of overloads. I guess we could take the union of all parameter types, subbing in `undefined` when a parameter isn't there.

**Comment by yortus** [comment]
Can I ask what's the use case/scenario for this? The only common scenario I know of where `next` is called with an argument is `co`-style async runners. In that case types passed to next are not related to each other even within the same iteration. They are only related to the types being yielded (not the other way around).

On the other hand, iterables for use in `for..of` and `Map` constructors and the like, they don't have anything passed to `next`.

**Comment by DanielRosenwasser** [comment]
This issue is largely to track user feedback on whether this is desired, or to at least point to some conversation about the idea. :smile:

**Comment by HerringtonDarkholme** [comment]
@yortus 
What about adding a phantom type parameter called `lastYielded`? TypeScript compiler can match the `lastYielded` value to choose what value type can be sent into yield expression.

```typescript
interface IterableIterator<T> {
  next<V, LastYielded>(sent: V): {value: T, done: boolean}
}

interface CoIterator extends IterableIterator<any> {
  next<T, Promise<T>>(sent: T): {value: any, done: boolean}
}
```

And in use:

```typescript

function *co(): CoIterator {
  let webpage = yield get('https;//typescriptlang.org') // get: (url: string) => Promise<string>
 // `LastYield` type is Promise<string>, so the `sent` value is `string`
}
```

I tried to express myself. If it is confusing, we can talk about this on gitter.

**Comment by mhegazy** [comment]
is not this the same as https://github.com/Microsoft/TypeScript/issues/2983?

**Comment by laughinghan** [comment]
## TL;DR: this is a duplicate of #2983 and should be closed

I believe this is indeed a duplicate of #2983, which discusses solving (among others) the problem where:

> - All `yield` expressions are typed as `any`, as you can see in the example comments below.

https://github.com/Microsoft/TypeScript/issues/2983#issuecomment-230414026

(I don't know what @mhegazy meant by "but for generators", #2983 is all about generators as far as I can tell.)
